awslabs/smithy-dafny
# feat(TestModels): Dependencies

*Description of changes:*

Changes overview:
* TestModels
  * Add `Dependencies` module
    * Add `UseLocalConstraintsService` to Smithy model defined [here](https://github.com/awslabs/polymorph/blob/smithy-test-models/TestModels/dependencies.smithy)
      * I noticed the `simpleConstraintsServiceReference` was untested when my first iteration of generated code for To/FromDafny for the service client called itself. This should have resulted in stack overflow, but since there were no tests exercising this path, this didn't fail. I added this to ensure the path is tested.
  * `ExtendableResource` module
    * Changed operation names from `XxResourceYy` to `XxExtendableResourceYy`; these conflicted with the `Resource` module when I imported them together
  * Makefiles
    * Update make targets to compile dependencies for polymorph dafny/net. (Dependency files were missing in GHA since GHA didn't compile them.)
* smithy-polymorph
  * Codegen for `FromDafny_CommonError`: Generate a single exception case per dependency. Use the "default" exception Dafny exception for the dependency.
  * Codegen for `ToDafny_CommonError`: Generate a `switch` statement based on the error namespace. For all errors in a given namespace, pass those errors to the namespace's corresponding `TypeConversion.ToDafny_CommonError`. This is done for all dependency namespaces.
  * Codegen for Shim `ConvertError`: Very similar to codegen for `ToDafny_CommonError`
  * Codegen for `To`/`FromDafny` for dependent services. These interact with the underlying impl to wrap/unwrap to convert to/from Dafny.

I had previously opened a few PRs for 1) Dependencies, 2) Handling wrapped errors from dependencies, and 3) Dependency local services.
I intended to use the smaller PRs to allow for more digestible changes, and have this PR be the work that is actually merged.
However, as part of combining the 3 changes, I learned I wrote bad logic for 2) and needed to rewrite it.
So, I don't think it's helpful to use the smaller PRs for understanding anymore. Sorry about that 😞 

* https://github.com/lucasmcdonald3/polymorph/pull/2
* https://github.com/lucasmcdonald3/polymorph/pull/3
* https://github.com/lucasmcdonald3/polymorph/pull/4

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
